operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go

```diff
@@ -0,0 +1,253 @@
+		Namespace: instance.Spec.Template.Namespace,
+		Name:      instance.Spec.Template.Name,
+	}
+	template := &crownlabsv1alpha2.Template{}
+
+	if err := r.Get(ctx, templateName, template); err != nil && errors.IsNotFound(err) {
+		// The declared template does not exist set the phase as failed and don't try again.
+		return false, fmt.Errorf("instance %s not found in namespace %s. It is not possible to complete the InstanceSnapshot %s",
+			instanceName.Name, instanceName.Namespace, isnap.Name)
+	} else if err != nil {
+		return true, fmt.Errorf("error in retrieving the template for InstanceSnapshot %s -> %w", isnap.Name, err)
+	}
+
+	// Retrieve the environment from the template.
+	var env *crownlabsv1alpha2.Environment = nil
+
+	if isnap.Spec.Environment.Name != "" {
+		for i := range template.Spec.EnvironmentList {
+			if template.Spec.EnvironmentList[i].Name == isnap.Spec.Environment.Name {
+				env = &template.Spec.EnvironmentList[i]
+				break
+			}
+		}
+
+		// Check if the specified environment was found.
+		if env == nil {
+			return false, fmt.Errorf("environment %s not found in template %s. It is not possible to complete the InstanceSnapshot %s",
+				isnap.Spec.Environment.Name, template.Name, isnap.Name)
+		}
+	} else {
+		// If the environment is not explicitly declared, take the first one.
+		env = &template.Spec.EnvironmentList[0]
+	}
+
+	// Check if the environment is a persistent VM.
+	if env.EnvironmentType != crownlabsv1alpha2.ClassVM || !env.Persistent {
+		return false, fmt.Errorf("environment %s is not a persistent VM. It is not possible to complete the InstanceSnapshot %s",
+			env.Name, isnap.Name)
+	}
+
+	// Check if the VM is running.
+	if instance.Spec.Running || instance.Status.Phase != "VmiOff" {
+		return false, fmt.Errorf("the vm is running. It is not possible to complete the InstanceSnapshot %s", isnap.Name)
+	}
+
+	return false, nil
+}
+
+// GetJobStatus sets a Job and returns its status.
+func (r *InstanceSnapshotReconciler) GetJobStatus(job *batch.Job) (bool, batch.JobConditionType) {
+	for _, c := range job.Status.Conditions {
+		// If the status corresponding to Success or failed is true, it means that the job completed.
+		if c.Status == corev1.ConditionTrue && (c.Type == batch.JobFailed || c.Type == batch.JobComplete) {
+			return true, c.Type
+		}
+	}
+
+	// Job did not complete.
+	return false, ""
+}
+
+// GetSnapshottingJob generates the job to be created.
+func (r *InstanceSnapshotReconciler) GetSnapshottingJob(isnap *crownlabsv1alpha2.InstanceSnapshot, job *batch.Job) {
+	var backoff int32 = 2
+	imagedir := isnap.Spec.Instance.Namespace
+	imagetag := fmt.Sprint(time.Now().Format("02012006t150405"))
+	volumename := isnap.Spec.Instance.Name
+
+	// Define volumes.
+
+	// Define VM VolumeSource.
+	persistentvolsrc := corev1.PersistentVolumeClaimVolumeSource{
+		ClaimName: volumename,
+	}
+	vmvolume := corev1.VolumeSource{
+		PersistentVolumeClaim: &persistentvolsrc,
+	}
```

Comment:
If you want those two can be merged together like (the same for the others):
```suggestion
	vmvolume := corev1.VolumeSource{
		PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
		         ClaimName: volumename,
	        },
	}
```